google/cloud/bigtable/batcher.py

```diff
@@ -24,6 +27,74 @@ class MaxMutationsError(ValueError):
     """The number of mutations for bulk request is too big."""
 
 
+class BatcherIsClosedError(ValueError):
+    """Batcher is already closed and not accepting any mutation."""
+
+
+class _MutationsBatchQueue(object):
+    """Private Threadsafe Queue to hold rows for batching."""
+
+    def __init__(self, max_row_bytes=MAX_ROW_BYTES, flush_count=FLUSH_COUNT):
+        """Specify the queue constraints"""
+        self._queue = queue.Queue(maxsize=flush_count)
+        self.total_mutation_count = 0
+        self.total_size = 0
+        self.max_row_bytes = max_row_bytes
+
+    def get(self, block=True, timeout=None):
+        """Retrieve an item from the queue. Recalculate queue size."""
+        row = self._queue.get(block=block, timeout=timeout)
+        mutation_size = row.get_mutations_size()
+        self.total_mutation_count -= len(row._get_mutations())
+        self.total_size -= mutation_size
+        return row
+
+    def put(self, item, block=True, timeout=None):
+        """Insert an item to the queue. Recalculate queue size."""
+
+        mutation_count = len(item._get_mutations())
+
+        if mutation_count > MAX_MUTATIONS:
+            raise MaxMutationsError(
+                "The row key {} exceeds the number of mutations {}.".format(
+                    item.row_key, mutation_count
+                )
+            )
```

Comment:
We should also block when the total inflight size is > 100MB. I think we'll need another variable FLUSH_SIZE to track when the current batch is full, and track the total size separately.